### PR TITLE
separate frontend and backend

### PR DESCRIPTION
**Pull request by Trolldemorted**
In the last few days I have been playing around to find a decent way to update an arbitrary number of MainPageViewModels from our background tasks.

This is not ready to be merged (as it does not support sending or opening additional MainPages yet), but the basic design should™ easily allow it:
`SignalLibHandle` is a singleton which acts as a gatekeeper between the frontend(s)/bg task and the library, and is enforcing stable ordering of all operations. New windows may register their dispatcher and their MPVM (`AddWindow(CoreDispatcher d, ISignalWindow w)`), and will be notified of any occuring event:

```
public interface ISignalWindow
{
    void AddOrUpdateConversation(SignalConversation conversation);
    void HandleIncomingMessage(SignalMessage message);
    void HandleIdentitykeyChange(LinkedList<SignalMessage> messages);
}
```

If the bg task is using the SignalLibHandle, it can notice that `MainWindowDispatcher` is null, and could call a bg-task-interface's methods.

What do you think so far?

### CODE REVIEW

**Comment by Trolldemorted** [comment]
A few nights without sleep later we can now open an arbitrary number of windows, whose states should™ be perfectly synchronized! 🎉 🎉 🎉 You can spawn a new window on a desktop by right-clicking on the taskbar icon and then left-clicking "Signal Private Messenger".

### things to test
- multiple windows, obviously
- Is the performance of the SignalLibHandle system sufficient on mobile devices?

### things to consider
- On which platforms do we want to enable the multi window feature?
- Should we provide more obvious/easy ways to spawn a new window? E.g. dragndrop a conversation?
- This most likely breaks on pages other than the MainPage since i did not find time to look at them so far. In which states/on what pages do we want to allow additional windows to be opened?
- What do we have to change so that the lib is easily usable by a bg process?
- I think I broke some mobile behaviour (which was previously fixed in e99e55c7f83bc5d870da502c8ad9cbcec8997c9d). What exactly was the problem with and happens during suspension?
- I broke group info requests. Will repair as soon as I find the time!

cc @golf1052 @SERVCUBED

File: Signal-Windows/App.xaml.cs [hunk 1]
```diff
@@ -44,24 +48,15 @@ sealed partial class App : Application
         public App()
         {
             SignalLogging.SetupLogging(true);
+            Logger.LogInformation("Signal-Windows App launching");
             this.InitializeComponent();
-            this.Suspending += OnSuspending;
-            this.Resuming += OnResuming;
```
Comment (golf1052) [inline Signal-Windows/App.xaml.cs]: Need to add this back for the MessagePipe to start up again on mobile.

File: Signal-Windows/App.xaml.cs [hunk 2]
```diff
@@ -188,40 +164,5 @@ private void OnNavigationFailed(object sender, NavigationFailedEventArgs e)
...
-                {
-                    await ViewModels.MainPageInstance.Shutdown();
-                }
-            }
-            Logger.LogInformation("OnSuspending() shutdown successful");
-            //TODO: Anwendungszustand speichern und alle Hintergrundaktivitäten beenden
-            deferral.Complete();
-        }
-
-        private async void OnResuming(object sender, object e)
```
Comment (golf1052) [inline Signal-Windows/App.xaml.cs]: As well as here.

**Comment by golf1052** [comment]
Just quickly looked over this. This is very similar to the refactor I did to support a background task. #108 is actually just 1 of 2 parts. Both parts were done, I just wanted to break it up into a manageable PR. Probably should have mentioned I also refactored the MessagePipe to not depend on the MainPageVM.

Definitely take a look at the background-thread branch because they do very similar things. The even model will work just as well, the only thing we would need to be careful of would be locking around things that should be locked like DB saves but this is already done.

**Comment by Trolldemorted** [comment]
Should we make the handle re-usable? Since it currently tracks our windows we might want to, as the view models call it frequently. This way we should be able to shutdown/startup our libsignal tasks w/o changing the frontend project at all.

Yeah I took a look at #108, that's why the branches are so similar :) How will the event model work, though? I didn't see any handling of dispatchers, and without the whole thing is gonna blow as soon as we want to update a secondary window.

**Comment by golf1052** [comment]
I'm not sure how the event based model fails with updating multiple views (which is where dispatchers come into play) but I guess this is due to my unfamiliarity with multiple views.

But I also may be unclear about what this change intends, it seems you want to support multiple independent windows instead of just the ability to spawn new views from one running app.

The one problem I see with supporting multiple independent windows is that we only want 1 instance of the app talking to Signal and updating the DB.

What I'm guessing happens with multiple views ([based on looking over this](https://docs.microsoft.com/en-us/windows/uwp/design/layout/show-multiple-views)) is that the view can be spawned and will I assume call OnNavigatedTo like the main view. I think multiple views makes sense for conversations, being able to pop out multiple conversation views while leaving the conversation thread picker in the main window makes sense. This also should still work with the event model, would need some more refactoring in the background-thread branch though.

When a conversation is selected the conversation view would subscribe to the relevant events. That conversation view would update it's view from fired events. When the conversation is deselected or the conversation view is closed then the events are unsubscribed.

I'm not currently seeing why the part that receives messages from Signal has to know about the view. When a message is received save it to the DB and then fire an event saying a new message is available. This way when there is no view everything still works the same. For the background task, instead of updating a view, we send out a notification instead.

**Comment by Trolldemorted** [comment]
Isn't supporting multiple windows and spawning new views in a new frame exactly the same?

Multiple windows are not meant only for single conversations. I use multiple virtual desktops, and it is very annoying that clicking on a notification throws me back to the one I opened Signal-Windows on.

Letting only one window talk with signal and update the db is not enough: We must ensure that all windows have the same state and display exactly the same content, so we must also synchronize displaying and creating messages too (if you send messages while receiving at the same time, we have to enforce the same order of messages in all windows). Thus both `display->save->send` and `receive->save->display` must be guarded by the same lock, and every display update in every window *must* be finished within the lock's protection, otherwise the ordering is racy (and very very bad things will happen. Our conversation view depends on knowing exactly how many messages are supposed to be there, allowing racy updates is not an option.

Dynamic registering of events makes things more complicated: If we don't want to be notified by literally everything, we have to keep track on what a window registered if we want to shut it down in a clean way. 

Also, we sometimes need immediate feedback from all windows: When a message comes in, we must know whether the corresponding thread is open, scrolled at the bottom and its window has focus before we can device if we want to dispatch a notification or not (TODO: use this to insert the message with the correct read state to reduce db usage).

[There are indeed ways that make events work with multiple windows](https://mtaulty.com/2015/09/16/m_15940/), but I am not convinced going that way is easier. The blog post is a great read on multiple windows, though! I used that post to get started with the multiple windows mindset.

Long story short, if we can guarantee that
- we can hold the handle lock while dispatching the display/newmessage events without running into deadlocks
- the viewmodels are not touched while an event handler is running (i.e. no await in the event handlers, or more locking)
- the event handlers are executed with the correct CoreDispatcher/SynchronizationContext
- we leak no memory by having stall registered event handlers
- the event handlers run parallel
- we can get feedback from them
- blocking for an event's completion with the handle lock held does not cause deadlocks

we can go the event way, but I fear that will require a lot more code and very precise proofreading and testing.

**Comment by golf1052** [comment]
I've given up on the event based version of this because events don't let us do these
- the event handlers run parallel
  - Possible but harder
- we can get feedback from them
  - Can't do

What is left on this PR before we can merge it in? I've gotten the background thread working in my branch so I can bring that into here.

**Comment by Trolldemorted** [comment]
my list of known issues:

1. the create new contact page doesn't work atm because it wants to talk to the signal server directly
2. messages may arrive before the first AddWindow call, so we won't display notifications for these.
3. frame.goback is creating new views and viewmodels (and iirc we cannot forbid that). Still not sure how we are gonna handle this - create an "interior force-cached" viewmodel and have the "real" viewmodels re-use them?
4. signallibhandle needs support for dispatching conversation changes (see the UpdateConversation skeleton), and the add contact and conversation settings pages needs to call it
5. if a new group arrives if multiple windows are open, a marshal exception arises. Looks like we bind to the same object, and I suspect the same will happen with new contacts:
```
[UI] [Warning] [Signal_Windows.Lib.IncomingMessages] HandleIncomingMessages() failed in line 0: One or more errors occurred. (Eine Schnittstelle, die für einen anderen Thread marshalled war, wurde von der Anwendung aufgerufen. (Exception from HRESULT: 0x8001010E (RPC_E_WRONG_THREAD)))
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout)
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks)
   at Signal_Windows.Lib.SignalLibHandle.DispatchAddOrUpdateConversation(SignalConversation conversation)
   at Signal_Windows.Lib.SignalLibHandle.SaveAndDispatchSignalMessage(SignalMessage message, SignalConversation conversation)
   at Signal_Windows.Lib.IncomingMessages.HandleSignalMessage(SignalServiceEnvelope envelope, SignalServiceContent content, SignalServiceDataMessage dataMessage, Boolean isSync, Int64 timestamp)
   at Signal_Windows.Lib.IncomingMessages.HandleMessage(SignalServiceEnvelope envelope)
   at Signal_Windows.Lib.IncomingMessages.OnMessage(SignalServiceMessagePipeMessage message)
   at libsignalservice.SignalServiceMessagePipe.ReadBlocking(IMessagePipeCallback callback)
   at Signal_Windows.Lib.IncomingMessages.HandleIncomingMessages()
```
6. this PR alone does not make Signal-Windows fit to allow db changes while it is loaded in memory. To be more specific, the MPVMs do not yet update their conversations in memory when resuming because so far they never had to. I propose we add something like `void HandleNewConversationList(List<SignalConversation>` to the ISignalWindow interface. This would also make our life easier if we want to register the first window in the SignalLibHandle's constructor call, which would make mitigating issue 2
7. we might want to rename some stuff. Most functions/classes got their names in a prototyping phase, so they are not the best. E.g. ISignalWindow should be something like ISignalFrontend or ISignalConsumer?
8. we might want to come up with something that allows Signal-Windows to interrupt the bg task. Imagine you didn't have internet for a longer period, so the background process is pulling down lots of messages, and the UI won't start because it cannot grab the SignalLibHandle because the bg task is holding it
9. [New windows are always spawned on the virtual desktop where the main (?) window is, and not on the virtual desktop the window was requested](https://stackoverflow.com/q/48179221/1569755)
10. We have to disable multiwindow for mobile

**Comment by Trolldemorted** [comment]
9. was fixed by 9881f3fa8b018890b6e2b69683d7fc7cdf490122!

Shall we merge and sort out the remaining issues piece by piece? That would avoid this merge from growing even bigger, but the master will not be production-ready.

**Comment by golf1052** [comment]
Yeah I feel like it would be better to just merge it in now. As long as it's in a mainly stable state.

**Comment by Trolldemorted** [comment]
It should be stable unless you click the add contact button :shipit:

Fixed some of the issues, will create one/some issues tomorrow, its 3am right now :(

**Comment by golf1052** [comment]
For the background thread to process incoming messages we're going to need to use events. Windows Runtime classes can only implement Windows Runtime interfaces. Meaning we could move the interface into the background thread project but that would be a weird hierarchy. Events however work fine.

**Comment by Trolldemorted** [comment]
What? Why? What about standard interfaces like IList? Can you push your current state so I can have a look?

**Comment by golf1052** [comment]
Check out the stuff in [MultiwindowTesting](https://github.com/golf1052/MultiwindowTesting). If you try to make RandomNumberPuller (a Runtime component) try to inherit INumber you'll get an error upon build. You can inherit from IRuntimeInterface fine however.

**Comment by Trolldemorted** [comment]
Ah, WinRT libs must only export WinRT Types. Whatever floats their boat, I think we can just use a not exported implementation of the interface like this:

```
internal class BGNumber : INumber
{
    public void Numbering()
    {
        throw new NotImplementedException();
    }
}
```

and have that class call our components' methods.
